hashintel/hash
# Improve type editing (introduce read-only mode, show type versions)

## 🌟 What is the purpose of this PR?

UX improvements in the type editor package and the type page in HASH:

1. Type editor implements `readonly` mode
2. HASH can handle versioned URIs for entity types (previously 404'd) – just redirects to latest for now
3. Show version numbers on types in dropdown, and in ontology chips
4. Link types in links table are clickable to go to their URI

## 🔗 Related links





- [Asana task for readonly mode](https://app.asana.com/0/1203179076056209/1203623179643234/f) _(internal)_
- [Slack thread of potential UX improvements](https://hashintel.slack.com/archives/C02LG39FJAU/p1677068435104359) _(internal)_

## 🚫 Blocked by




- [x] Merging of #2113 

## 🔍 What does this change?

- Add a readonly context inside the type editor and use it to disable various hover effects and fields
- Update type editor route in HASH so that it captures versioned URIs
- Allow passing custom routing into type-editor (so that we can use Next routing on links and not require full page navigation)
- Update `OntologyChip` so that it doesn't strip out version information
- Remove no longer unnecessary type assertions

## 📜 Does this require a change to the docs?



- The docs for x need updating to explain that y

## ⚠️ Known issues



## 🐾 Next steps

**Type editor updates**

1. Enable updating link types
5. Fix updating property types (clicking the button does not load the new property type in)
6. Only show the latest version of an entity type in the dorpdown
7. Publish new version

**blockprotocol.org updates**
1. Add new version of type editor
2. Show that a link entity is a link entity
3. Allow switching an entity to a link entity and vice versa

## ❓ How to test this?



1.  Try a type that should be readonly (not yours), e.g. `/@system-user/types/entity-type/user/v/1`, check it's readonly
2. Try a type that should work as normally (one of yours), check it's editable
3. Try different version URLs, should go to latest

## 📹 Demo

https://user-images.githubusercontent.com/37743469/220891459-72a506fe-9d86-4dc2-a7ea-dec24e65cb91.mp4
